cmd/ncproxy/server.go

```diff
@@ -88,3 +119,59 @@ func (s *server) serve(ctx context.Context, ttrpcListener net.Listener, grpcList
 		serveErr <- trapClosedConnErr(s.grpc.Serve(grpcListener))
 	}()
 }
+
+// reconnectComputeAgents creates new compute agent connections from the database of
+// active compute agent addresses and adds them to the compute agent client cache
+// this MUST be called before the server start serving anything so that we can
+// ensure that the cache is ready when they do.
+func reconnectComputeAgents(ctx context.Context, agentStore *computeAgentStore, agentCache *computeAgentCache) {
+	computeAgentMap, err := agentStore.getActiveComputeAgents(ctx)
+	if err != nil && errors.Is(err, errBucketNotFound) {
+		// no entries in the database yet, return early
+		log.G(ctx).WithError(err).Debug("no entries in database")
+		return
+	} else if err != nil {
+		log.G(ctx).WithError(err).Error("failed to get compute agent information")
```

Comment:
Is it safe to ignore the error and allow the service to start? Could the database be in a state that would break things later on? What are the implications if some agents are not reconnected?

Reply:
The possible errors returned from that call are:

1. ErrTxNotWritable: is returned when performing a write operation on a read-only transaction 
    This won't happen since the code in `getActiveComputeAgents` never attempts to write to the transaction. 

2. ErrDatabaseReadOnly: is returned when a mutating transaction is started on a read-only database
    This won't happen since the `getActiveComputeAgents` call does not attempt to write to the database. The database is also opened with rw, so this can't happen anyways. 

3. ErrTxClosed: is returned when committing or rolling back a transaction that has already been committed or rolled back.
    This could happen if the database is nil, which it can't be at the point, or if the above happens. But since the transaction is created within the call to `View` and we never attempt to do any of the above manually, I think we're pretty safe from this. 
    If this happens in some weird case, we would fail to reconnect to the compute agents. This would cause calls from the compute agent or from ncproxy to the compute agent to fail. For calls from the compute agent, the only times this could happen would be when a call is made to setup or teardown the pod's network. In the setup case, not connecting would just cause us to fail to start the pod. This is fine because we can just try again. In the teardown case, the call would not prevent us from shutting down the pod, but it would cause us to fail to request the node network service to tear down the network. In that case, we can't really do anything since we don't have any information on what containers exist to be able to try to request the node network service to tear down their networking. It would need to be the vnetagent's responsibility to clean up any host resources that are laying around. 
   For calls from ncproxy to the compute agent, we would fail to connect to the compute agent and just return the error. The node network service could decide at this point to teardown the networking setup. 
 
4. ErrDatabaseNotOpen: is returned when the database has not been opened 
    This won't happen since we open the database before the call to reconnect the compute agents.

4. errBucketNotFound: we return this in the above call when we can't find the bucket in question.  
    If this happens, it just means that the database is new and has no content, in that case, the reconnect function will return after doing nothing and we continue on as normal.

Reply:
As for if only some of the compute agents are not connected, we how have a similar situation where the node network service would need to own cleaning up host resources that are laying around. But this would remove the entry in the database, so we would be safe from old entries polluting in that case.

Reply:
Some of this info may be applicable to the reconnect overview suggested in a previous comment.